paasta_tools/contrib/check_orphans.py

```diff
@@ -71,19 +74,52 @@ def get_zk_data(blacklisted_services: Set[str]) -> SmartstackData:
 
 
 class InstanceTuple(NamedTuple):
+    # paasta_host may be different from the service's host if running on k8s.
+    # We need the actual PaaSTA host because the k8s pod does not listen for
+    # xinetd connections.
+    paasta_host: str
     host: str
     port: int
     service: str
 
 
 def read_from_zk_data(registrations: SmartstackData) -> Set[InstanceTuple]:
     return {
-        InstanceTuple(instance_data["host"], instance_data["port"], service)
+        InstanceTuple(
+            host_to_ip(instance_data["name"], instance_data["host"]),
+            instance_data["host"],
+            instance_data["port"],
+            service,
+        )
         for service, instance in registrations.items()
         for instance_data in instance.values()
     }
 
 
+@functools.lru_cache()
+def host_to_ip(host, fallback):
+    """Try to resolve a host to an IP with a fallback.
+
+    Because DNS resolution is relatively slow and can't be easily performed
+    using asyncio, we cheat a little and use a regex for well-formed hostnames
+    to try to guess the IP without doing real resolution.
+
+    A fallback is needed because in some cases the nerve registration does not
+    match an actual hostname (e.g. "prod-db15" or "prod-splunk-master").
+    """
+    for match in (
+        re.match(r"^(\d+)-(\d+)-(\d+)-(\d+)-", host),
+        re.match(r"^ip-(\d+)-(\d+)-(\d+)-(\d+)$", host),
+    ):
+        if match:
```

Comment:
Maybe replace it with something like
```
# somewhere above
_RE_HOSTNAMES = re.compile(r"^(?:ip-)?(\d+)-(\d+)-(\d+)-(\d+)")
...
# here
if _RE_HOSTNAMES.match(host):
```
?

Reply:
I can move it if you think it improves readability? I am not sure moving the regex far from the function that uses it does improve readability, and the performance should be the same anyway.

Reply:
I especially dislike that there are two regexes when one is more than enough 🙂 
Performance will be better since here you're compiling regexes on every function call, while by using a pre-compiled regex you'll do that only once.

Reply:
I don't feel strongly about this but I feel like this is playing regex golf at the expense of readability; these are targeting two different hostname formats and I think using two regexes is appropriate. After your proposed change it's not as obvious what the two hostname formats are.

The pre-compiling shouldn't matter due to the internal regex cache: https://docs.python.org/3/library/re.html#re.compile

If you still think it is worth doing I will do it though.

Reply:
> The pre-compiling shouldn't matter due to the internal regex cache: https://docs.python.org/3/library/re.html#re.compile

good to know

ok, let's resolve it then.